src/Skoruba.MultiTenant.Identity/Validators/UserRequiresTwoFactorAuthenticationValidator.cs

```diff
@@ -0,0 +1,60 @@
+﻿using Microsoft.AspNetCore.Identity;
+using Skoruba.MultiTenant.Abstractions;
+using Skoruba.MultiTenant.Configuration;
+using System.Threading.Tasks;
+
+namespace Skoruba.MultiTenant.Identity.Validators
+{
+    public class UserRequiresTwoFactorAuthenticationValidator<TUser> : IUserValidator<TUser>
+        where TUser : class
+    {
+        private readonly ISkorubaTenantContext _skorubaTenantContext;
+
+        public UserRequiresTwoFactorAuthenticationValidator(ISkorubaTenantContext skorubaTenantContext)
+        {
+            _skorubaTenantContext = skorubaTenantContext;
+        }
+
+        public Task<IdentityResult> ValidateAsync(UserManager<TUser> manager, TUser user)
+        {
+            if (user is IdentityUser)
+            {
+                IdentityUser identityUser = user as IdentityUser;
+
+                if (identityUser.TwoFactorEnabled)
+                {
+                    return Task.FromResult(IdentityResult.Success);
+                }
+
+                if (_skorubaTenantContext.TenantResolutionRequired && !_skorubaTenantContext.TenantResolved)
+                {
+                    throw MultiTenantException.MissingTenant;
```

Comment:
The reason i'm throwing an exception is because **it should not happen**.  If it does not resolve and is required then there is something wrong with the multi tenant implementation.  I worry that a graceful exit might hide a larger problem.

Reply:
Agreed.